File: backend/app/routers/coordinators.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.core.db import get_db
from app.core.deps import require_role
from app.core.roles import Role
from app.models.user import User
from app.schemas.user import AvailabilityUpdate, UserOut
from app.services.assignment import events_needing_reassignment, reassign_event

router = APIRouter(prefix="/coordinators", tags=["coordinators"])


@router.patch("/me/availability", response_model=UserOut)
def set_my_availability(
    body: AvailabilityUpdate,
    db: Session = Depends(get_db),
    caller: User = Depends(require_role(Role.COORDINATOR)),
) -> UserOut:
    """Toggle the caller's own availability.

    `require_role` only proves identity from the JWT (id + role); the row
    itself -- and the name a reassignment note needs -- still has to come
    from the database, the same way GET /auth/me re-fetches its profile.

    Turning availability OFF immediately reassigns every event currently
    active under this Coordinator to another available one (or leaves it
    unassigned if none exists) -- the whole point of the story is that an
    Organiser's event does not sit stuck with a Coordinator who cannot work
    on it.
    """
    user = db.get(User, caller.id)
    if user is None:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Could not validate credentials")

    if user.is_available != body.is_available:
        user.is_available = body.is_available
        if body.is_available is False:
            for event in events_needing_reassignment(db, user.id):
                reassign_event(db, event, outgoing=user)

    db.commit()
    db.refresh(user)
    return UserOut.model_validate(user)

File: frontend/src/lib/coordinators.ts
// Mirror of backend/app/routers/coordinators.py and schemas/user.py --
// keep field names byte-identical.

import { apiFetch } from './api'

export interface CoordinatorSelf {
  id: number
  name: string
  email: string
  role: string
  is_available: boolean
  created_at: string
}

/** Toggle the signed-in Coordinator's own availability.
 *
 *  Turning it off reassigns every event currently active under them to
 *  another available Coordinator server-side -- see the "Mark myself
 *  unavailable" story. There is nothing else to call here to make that
 *  happen; it is a side effect of this one request. */
export function setMyAvailability(isAvailable: boolean): Promise<CoordinatorSelf> {
  return apiFetch('/coordinators/me/availability', {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ is_available: isAvailable }),
  }) as Promise<CoordinatorSelf>
}
